keps/sig-apps/3329-retriable-and-non-retriable-failures/README.md

```diff
@@ -0,0 +1,1543 @@
+- PreemptedByScheduler (Pod preempted by kube-scheduler)
+- EvictedByTaints (Pod evicted by kube-controller-manager due to taints)
+- GCed (an orphaned Pod terminated by GC)
+- EvictedByAPI (Pod evicted by Eviction API)
+
+The already existing `status.conditions` field in Pod will be used by kubernetes
+control plane components (kube-scheduler and kube-controller-manager) to append
+a dedicated condition when they send the delete operation.
+
+The API call to append the condition will be issued as a pod status update call
+before the Pod delete request. This way the Job controller will already see the
+condition when handling a failed pod.
+
+During the implementation process we are going to review the places where the
+pod delete requests are issued to modify the code to also append a meaningful
+condition with dedicated `Type`, `Reason` and `Message` fields based on the
+invocation context.
+
+### JobSpec API
+
+We extend the Job API in order to allow to apply different actions depending
+on the conditions associated with the pod failure.
+
+```golang
+// BackoffPolicyAction specifies how a Pod failure is handled.
+type BackoffPolicyAction string
+
+const (
+
+	// This is an action which might be taken on pod failure - mark the
+	// pod's job as Failed and terminate all running pods.
+	BackoffActionTerminate BackoffPolicyAction = "Terminate"
+
+	// This is an action which might be taken on pod failure - the pod will be
+	// restarted and the counter for .backoffLimit will not be incremented.
+	BackoffActionIgnore BackoffPolicyAction = "Ignore"
+)
+
+type BackoffPolicyOnExitCodesOperator string
+
+const (
+	BackoffPolicyOnExitCodesOpIn    BackoffPolicyOnExitCodesOperator = "In"
+	BackoffPolicyOnExitCodesOpNotIn BackoffPolicyOnExitCodesOperator = "NotIn"
+)
+
+type BackoffPolicyOnExitCodesRequirement struct {
+	// Restricts the check for exit codes to only apply to the container with
+	// the specified name. When empty the rule applies to all containers
+	// +optional
+	ContainerName *string
+
+	// Represents the relationship between the container exit code(s) and the
+	// specified values.
+	Operator BackoffPolicyOnExitCodesOperator
+
+	// Specifies the set of values. Each returned container exit code (might be
+	// multiple in case of multiple containers) is checked against this set of
+	// values with respect to the operator
+	Values []int
+}
+
+type BackoffPolicyOnPodConditionsOperator string
+
+const (
+	BackoffPolicyOnPodConditionsOpIn BackoffPolicyOnPodConditionsOperator = "In"
+)
+
+type BackoffPolicyOnPodConditionsRequirement struct {
+
+	// Represents the relationship between the actual Pod condition types
+	// and the set of specified Pod condition types
+	Operator BackoffPolicyOnPodConditionsOperator
+
+	// Specifies the set of values. Each actual pod condition type is checked
+	// against this set of values with respect to the operator
+	Values []api.PodConditionType
```

Comment:
I would name this types, for clarity. Also you need to say in the documentation that it only matches conditions with Status=True

Reply:
Added a note on the status=True.

I like Values as it is more similar to the "onExitCodes" API. Maybe instead we say "onPodConditionTypes" level up?

Reply:
that might work

Reply:
We can leave this kind of low level details to code review.